statistics/handle/cache/internal/lfu/lfu_cache.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lfu
+
+import (
+	"sync/atomic"
+
+	"github.com/dgraph-io/ristretto"
+	"github.com/pingcap/tidb/statistics"
+	"github.com/pingcap/tidb/statistics/handle/cache/internal"
+	"github.com/pingcap/tidb/util/intest"
+)
+
+// LFU is a LFU based on the ristretto.Cache
+type LFU struct {
+	cache        *ristretto.Cache
+	resultKeySet *keySetShard
+	cost         atomic.Int64
+}
+
+// NewLFU creates a new LFU cache.
+func NewLFU(totalMemCost int64) (*LFU, error) {
+	result := &LFU{}
+	bufferItems := int64(64)
+	if intest.InTest {
+		bufferItems = 1
+	}
+	cache, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters:        1024 * 1024 * 1024, // TODO(hawkingrei): make it configurable for NumCounters and MaxCost
+		MaxCost:            totalMemCost,
+		BufferItems:        bufferItems,
+		OnEvict:            result.onEvict,
+		OnExit:             result.onExit,
+		IgnoreInternalCost: intest.InTest,
+		Metrics:            intest.InTest,
+	})
+	result.cache = cache
+	result.resultKeySet = newKeySetShard()
+	return result, err
+}
+
+// Get implements statsCacheInner
+func (s *LFU) Get(tid int64, _ bool) (*statistics.Table, bool) {
+	result, ok := s.cache.Get(tid)
+	if !ok {
+		return nil, ok
+	}
+	return result.(*statistics.Table), ok
+}
+
+// Put implements statsCacheInner
+func (s *LFU) Put(tblID int64, tbl *statistics.Table, _ bool) {
+	ok := s.cache.Set(tblID, tbl, tbl.MemoryUsage().TotalTrackingMemUsage())
+	if ok { // NOTE: `s.cache` and `s.resultKeySet` may be inconsistent since the update operation is not atomic, but it's acceptable for our scenario
+		s.resultKeySet.Add(tblID)
+		s.cost.Add(tbl.MemoryUsage().TotalTrackingMemUsage())
```

Comment:
Will the cost be accumulated repeatedly?

Reply:
No, because if the operator is ```update```, although this will add the new object's memory, OnExit will be called in the background and delete the old object's memory.